### ISSUE TITLE
salt-ssh sudo issues

### ISSUE THREAD

**Issue by streetmapp** [saltstack/salt#25134]
Having an issue with salt-ssh using sudo where `requiretty` is enabled for sudo use.

`test.ping` works:

`$ salt-ssh '*' --config-dir=/home/tmapp/tomcat/salt/conf/ test.ping                
test1:
    True`

Passing a raw command works:

```
$ salt-ssh '*' --config-dir=/home/tmapp/tomcat/salt/conf/ -r 'ls -l /tmp/tylertest' 

test1:
    ----------
    retcode:
        0
    stderr:
        Connection to 10.24.222.22 closed.
    stdout:
        -rw-r--r-- 1 tmapp wheel 0 Jul  2 14:43 /tmp/tylertest
```

But trying to use a module does not:

```
$ salt-ssh '*' --config-dir=/home/tmapp/tomcat/salt/conf/ file.exists /tmp/tylertest
test1:
    ----------
    retcode:
        0
    stderr:
        Connection to 10.24.222.22 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
```

```
$ salt-ssh --versions-report
           Salt: 2015.5.2
         Python: 2.7.9 (default, Jul  2 2015, 13:01:57)
         Jinja2: 2.7.3
       M2Crypto: Not Installed
   msgpack-python: 0.4.6
   msgpack-pure: 0.1.3
       pycrypto: 2.6.1
        libnacl: 1.4.3
         PyYAML: 3.11
          ioflo: 1.3.6
          PyZMQ: 14.7.0
           RAET: 0.6.3
            ZMQ: 4.1.2
           Mako: 1.0.1

```

Based on similar issues surrounding `requiretty`, thought the version we are trying to use would not have these issues in this version of salt-ssh.

If there are other details needed, let me know and I will work to provide those.

**Comment by jfindlay** [comment]
@streetmapp, thanks for the report.  Will you also post the output of `salt-ssh '*' -l debug --config-dir=/home/tmapp/tomcat/salt/conf/ file.exists /tmp/tylertest`?

**Comment by streetmapp** [comment]
@jfindlay, here you go: 

```
$ salt-ssh '*' -l debug --config-dir=/home/tmapp/tomcat/salt/conf/ file.exists /tmp/tylertest
[INFO    ] Loading Saltfile from '/home/tmapp/tomcat/salt/conf/Saltfile'
[DEBUG   ] Reading configuration from /home/tmapp/tomcat/salt/conf/Saltfile
[DEBUG   ] Reading configuration from /home/tmapp/tomcat/salt/conf/master
[DEBUG   ] Configuration file path: /home/tmapp/tomcat/salt/conf/master
[DEBUG   ] LazyLoaded flat.targets
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Jinja search path: ['/home/tmapp/tomcat/salt/var/cache/salt/master/files/base']
[DEBUG   ] Rendered data from file: /home/tmapp/tomcat/salt/conf/roster:
test1:
  host: 10.24.222.22
  user: tmapp
  sudo: True
  tty: True

[DEBUG   ] Results of YAML rendering: 
OrderedDict([('test1', OrderedDict([('host', '10.24.222.22'), ('user', 'tmapp'), ('sudo', True), ('tty', True)]))])
[INFO    ] minions list: {'test1': OrderedDict([('host', '10.24.222.22'), ('user', 'tmapp'), ('sudo', True), ('tty', True)])}
[DEBUG   ] LazyLoaded roots.envs
[DEBUG   ] Could not LazyLoad roots.init
[DEBUG   ] Updating roots fileserver cache
[DEBUG   ] LazyLoaded local_cache.prep_jid
[DEBUG   ] Could not LazyLoad file.exists
[DEBUG   ] Performing shimmed, blocking command as follows:
file.exists /tmp/tylertest
[DEBUG   ] Executing command: scp   -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  /home/tmapp/tomcat/salt/var/cache/salt/master/ssh_shim/shim_EWmb_I 10.24.222.22:.9f1decfea59e
[DEBUG   ] Child Forked! PID: 10981  STDOUT_FD: 7  STDERR_FD: 9
[DEBUG   ] Terminal Command: /bin/sh -c scp   -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  /home/tmapp/tomcat/salt/var/cache/salt/master/ssh_shim/shim_EWmb_I 10.24.222.22:.9f1decfea59e
[DEBUG   ] Executing command: ssh 10.24.222.22 -t -t -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  /bin/sh $HOME/.9f1decfea59e
[DEBUG   ] Child Forked! PID: 10983  STDOUT_FD: 7  STDERR_FD: 9
[DEBUG   ] Terminal Command: /bin/sh -c ssh 10.24.222.22 -t -t -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  /bin/sh $HOME/.9f1decfea59e
[DEBUG   ] Executing command: ssh 10.24.222.22 -t -t -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  rm $HOME/.9f1decfea59e
[DEBUG   ] Child Forked! PID: 10984  STDOUT_FD: 7  STDERR_FD: 9
[DEBUG   ] Terminal Command: /bin/sh -c ssh 10.24.222.22 -t -t -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=65 -o Port=22 -o IdentityFile=/home/tmapp/tomcat/salt/ssh/salt-ssh.rsa -o User=tmapp  rm $HOME/.9f1decfea59e
[DEBUG   ] RETCODE 10.24.222.22: 0
[DEBUG   ] SHIM retcode(0) and command: _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
[DEBUG   ] LazyLoaded nested.output
test1:
    ----------
    retcode:
        0
    stderr:
        Connection to 10.24.222.22 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
```

**Comment by streetmapp** [comment]
In further reading, we have noticed that issue #24863 is also output that we have seen in testing.

**Comment by jfindlay** [comment]
@streetmapp, thanks for the extra info.

**Comment by jfindlay** [comment]
I am able to confirm this, except that some execution modules work and others do not.  (So far all that I've tried of `test.*` work.)

### configs
```yaml
jmoney-main salt 2015.8 # cat /etc/salt/roster
jmoney-rhel-6: {host: 54.205.70.12, priv: /root/.ssh/id_EC2, sudo: true, tty: true,
  user: ec2-user}
jmoney-rhel-7.1: {host: 54.204.151.118, priv: /root/.ssh/id_EC2, sudo: true, tty: true,
  user: ec2-user}
jmoney-rhel-7.2: {host: 54.205.182.166, priv: /root/.ssh/id_EC2, sudo: true, tty: true,
  user: ec2-user}
```

### commands
```yaml
jmoney-main salt 2015.8 # salt-ssh 'jmoney-rhel-*' test.ping
jmoney-rhel-7.2:
    True
jmoney-rhel-7.1:
    True
jmoney-rhel-6:
    True
```
```yaml
jmoney-main salt 2015.8 # salt-ssh 'jmoney-rhel-*' test.fib 5
jmoney-rhel-7.2:
    - 5
    - 4.05311584473e-06
jmoney-rhel-7.1:
    - 5
    - 5.96046447754e-06
jmoney-rhel-6:
    - 5
    - 5.96046447754e-06
```
```yaml
jmoney-main salt 2015.8 # salt-ssh 'jmoney-rhel-*' file.touch /tmp/test
jmoney-rhel-7.2:
    True
jmoney-rhel-7.1:
    True
jmoney-rhel-6:
    True
```

Running a nonexistent exec module function (`file.exists`) does not produce the appropriate error message:
```yaml
jmoney-main salt 2015.8 # salt-ssh 'jmoney-rhel-*' file.exists /tmp/test
jmoney-rhel-7.2:
    ----------
    retcode:
        0
    stderr:
        Connection to 54.205.182.166 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
jmoney-rhel-7.1:
    ----------
    retcode:
        0
    stderr:
        Connection to 54.204.151.118 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
jmoney-rhel-6:
    ----------
    retcode:
        0
    stderr:
        Connection to 54.205.70.12 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
```

Also, `state.apply`, `state.sls`, `state.highstate` fail by showing the whole salt-ssh config:
```yaml
jmoney-main salt 2015.8 # cat /srv/salt/test.sls
try:
  cmd.run:
    - name: ls -lh /tmp/test
```
```yaml
jmoney-main salt 2015.8 # salt-ssh 'jmoney-rhel-*' state.apply test
jmoney-rhel-7.2:
    ----------
    __master_opts__:
        ----------
        [...]
    _caller_cachedir:
        /var/cache/salt/master
    _error:
        Failed to return clean data
    _ssh_version:
        6.7p1 Debian-5
    ext_pillar:
        |_
          ----------
          git:
              master git://github.com/jfindlay/salt-pillar.git
    extension_modules:
        /var/cache/salt/extmods
    file_roots:
        ----------
        base:
            - /root/salt-pkg-tests
            - /root/precipitate
            - /srv/salt
    id:
        jmoney-rhel-7.2
    pillar_roots:
        ----------
        base:
            - /srv/pillar
            - /srv/spm/pillar
    retcode:
        0
    stderr:
        Connection to 54.205.182.166 closed.
    stdout:
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
        {
            "local": {
                "fun_args": [],
                "jid": "20160120170507837353",
                "return": {
                    [...]
                },
                "retcode": 0,
                "fun": "test.opts_pkg",
                "id": "jmoney-rhel-7.2"
            }
        }
```

### versions
```yaml
jmoney-main salt 2015.8 # salt-ssh --versions
Salt Version:
           Salt: 2015.8.3-238-ge425cbd

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: debian 8.2
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.2
```

**Comment by abyth** [comment]
Any news on this issue? Fix  #35211, which was cross-referenced, seems to not resolve this issue
